tests/unit/accounts/test_views.py

```diff
@@ -847,39 +911,49 @@ def test_get_returns_recovery_code_status(self, pyramid_request, redirect_url):
     def test_totp_auth(
         self,
         monkeypatch,
-        pyramid_request,
+        db_request,
         redirect_url,
         has_recovery_codes,
         remember_device,
+        make_email_renderers,
+        metrics,
     ):
+        make_email_renderers("unrecognized-login")
         remember = pretend.call_recorder(lambda request, user_id: [("foo", "bar")])
         monkeypatch.setattr(views, "remember", remember)
 
         _remember_device = pretend.call_recorder(lambda *a, **kw: None)
         monkeypatch.setattr(views, "_remember_device", _remember_device)
 
-        query_params = {"userid": str(1)}
+        user = UserFactory.create(
+            with_verified_primary_email=True,
+            username="testuser",
+            name="Test User",
+            last_login=(
+                datetime.datetime.now(datetime.UTC) - datetime.timedelta(days=1)
+            ),
+        )
+        monkeypatch.setattr(
+            type(user),
+            "has_recovery_codes",
+            property(lambda u: has_recovery_codes),
+        )
+        user.record_event = pretend.call_recorder(lambda *a, **kw: None)
```

Comment:
thought: instead of monkeypatching, we could pass something like:
```python
recovery_codes=[RecoveryCode(code="fake")] if has_recovery_codes else [],
```
to UserFactory.create() - but since you're following an existing pattern, that's fine to defer.